cylc/flow/etc/cylc

```diff
@@ -119,77 +95,89 @@ CYLC_HOME_ROOT="${CYLC_HOME_ROOT:-/opt}"
-#                           ║        set       ║ not set ║
-# ╔═══════════════╦═════════╬══════════════════╩═════════╣
-# ║ CYLC_ENV_NAME ║   set   ║      Use CYLC_ENV_NAME     ║
-# ║               ╠═════════╬══════════════════╦═════════╣
-# ║               ║ not set ║ Use CYLC_VERSION ║  "cylc" ║
-# ╚═══════════════╩═════════╩══════════════════╩═════════╝
 if [[ -z "${CYLC_HOME}" ]]; then
     if [[ -z "${CYLC_ENV_NAME}" ]]; then
         if [[ -n "${CYLC_VERSION}" ]]; then
             CYLC_ENV_NAME="cylc-$CYLC_VERSION"
         else
-            # Default version symlink - export CYLC_ENV_NAME to ensure it
-            # does not get overridden by CYLC_VERSION
-            export CYLC_ENV_NAME="cylc"
+            # Use default version (symlink)
+            CYLC_ENV_NAME="cylc"
         fi
     fi
+    # CYLC_VERSION gets set to the actual version of Cylc used within workflows.
+    # Therefore we export CYLC_ENV_NAME to ensure it gets used rather than
+    # CYLC_VERSION in subsequent calls to the wrapper.
+    export CYLC_ENV_NAME
     for ROOT in "${CYLC_HOME_ROOT}" "${CYLC_HOME_ROOT_ALT}"; do
         if [[ -d "${ROOT}/${CYLC_ENV_NAME}" ]]; then
             CYLC_HOME="${ROOT}/${CYLC_ENV_NAME}"
+            # If CYLC_HOME is a symlink then replace it with the real path and
+            # set CYLC_ENV_NAME to the linked environment to ensure changes to
+            # symlinked environments can't affect running workflows.
+            if [[ -L ${CYLC_HOME} ]]; then
+                CYLC_HOME=$(readlink -f "${CYLC_HOME}")
+                CYLC_ENV_NAME=${CYLC_HOME##*/}
+            fi
             break
         fi
     done
+    if [[ -z "${CYLC_HOME}" ]]; then
+        MSG="ERROR: $CYLC_ENV_NAME not found in $CYLC_HOME_ROOT"
+        if [[ -n "${CYLC_HOME_ROOT_ALT}" ]]; then
+            MSG="${MSG} or ${CYLC_HOME_ROOT_ALT}"
+        fi
+        echo 1>&2 "$MSG"
+        exit 1
+    fi
 fi
-if [[ -z "${CYLC_HOME}" ]]; then
-    MSG="ERROR: $CYLC_ENV_NAME not found in $CYLC_HOME_ROOT"
-    if [[ -n "${CYLC_HOME_ROOT_ALT}" ]]; then
-        MSG="${MSG} or ${CYLC_HOME_ROOT_ALT}"
+
+# Legacy support for Rose
+if [[ ${0##*/} =~ ^ros ]]; then
+    # Prior to Cylc 8, Rose used a standalone installation
+    if [[ -n "${CYLC_ENV_NAME}" ]]; then
+        ROSE_HOME_ROOT="${ROSE_HOME_ROOT:-$CYLC_HOME_ROOT}"
+        if [[ ${CYLC_ENV_NAME:-} =~ ^cylc-7 ]]; then
+            # Cylc 7: Use ROSE_HOME / ROSE_VERSION to select the installation
+            if [[ -z "${ROSE_HOME:-}" ]]; then
+                if [[ -n "${ROSE_VERSION:-}" ]]; then
+                    CYLC_HOME="${ROSE_HOME_ROOT}/rose-${ROSE_VERSION}"
+                else
+                    # Use default version (symlink)
+                    CYLC_HOME="${ROSE_HOME_ROOT}/rose"
+                fi
+            else
+                CYLC_HOME="${ROSE_HOME}"
+            fi
+        elif [[ ${1:-} == "edit" || ${1:-} == "config-edit" ]]; then
```

Comment:
I think it would be reasonable to add a warning - consider the following silly, but plausible scenario:

```bash
> export CYLC_VERSION=8.0b1; export ROSE_VERSION=2019.01.1;   rose version
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'metomi.rose' has no attribute '__version__'
```

Reply:
This isn't related to this change - you can't use the wrapper with Cylc 8 for anything prior to rc1 because we no longer set PATH so the old rose bash cli doesn't work.